### ISSUE TITLE
Where I should request a new Extension?

### ISSUE THREAD

**Issue by rilian-la-te** [flatpak/flatpak#1708]
## Linux distribution and version
Any.
## Flatpak version
Any.
## Description of the problem
I want to write a Flatpak runtime extension for Flatpak apps to give them ability to integrate with system Global Menu. 
In this extension I will place appmenu-gtk-module, dbusmenu-glib, and JAyatana if allowed. And this extension should also contain com.canonical.AppMenu.Registrar, and allow all applications to contact to it.
## Steps to reproduce
It is a question, not a bug. Just try to use any application with menu via Flatpak when global menu installed. And see difference (menu staying in app).

**Comment by rilian-la-te** [comment]
How I should request this? Or I can just place extension on GitHub and it will be turned on after compile and install?

**Comment by TingPing** [comment]
Here is a fine place to request it. The question is just do we want a GTK module extension point. They are kinda deprecated (removed in GTK4) and ones that dig into the application like `appmenu-gtk-module` is probably questionable.

**Comment by rilian-la-te** [comment]
AFAIK, only IM modules removed in GTK4, not all. But I can assume wrong about this.

But this is not only for GTK, it is for Qt and all other toolkits. Just extension point to provide some features for desktop, not necessary a Global Menu.

**Comment by TingPing** [comment]
> But this is not only for GTK, it is for Qt and all other toolkits. Just extension point to provide some features for desktop

Well there would have to be a different extension point for each toolkit, that is the only logical way.

It is worth noting that Gtk already has global menu support built in, via gmenu, applications and shells just need to use it (latest Plasma does I believe, I'm sure some gnome-shell extension does too).

**Comment by rilian-la-te** [comment]
GTK yes, but if you use GMenuModel. If you use GTK Menubar, it is no support. And you need to use my module to achieve it.

**Comment by alexander255** [comment]
Bump. How can we distribute the appmenu-gtk-module for Flatpak?

**Comment by matthiasclasen** [comment]
bundle it with the app

**Comment by rilian-la-te** [comment]
@matthiasclasen this is not a good point. We should able to add appmenu-gtk-module to all Flatpak apps, if we want. I think extension is more suitable for this.

**Comment by TingPing** [comment]
My original comments are still relevant. That module is kind of a hack and GTK supports exported menus natively.

Anyway honestly if you just add the extension point to the fdo runtime in a MR it can be discussed there.

**Comment by alexander255** [comment]
@TingPing: GMenuModel is a horrible „replacement“ for GtkMenuBar (unless you're a GNOME dev maybe), so please don't mix up the two. This issue is about being able to export GtkMenuBar using a GTK module.

**Comment by rilian-la-te** [comment]
@TingPing GMenuModel has a way less features than GtkMenuBar. Yes, appmenu-gtk-module translates between the two as hack, but it allows more than writing model natively.

**Comment by rilian-la-te** [comment]
@TingPing where is a fdo runtime repo? I can try to make appmenu-gtk-module runtime extension)

**Comment by TingPing** [comment]
https://gitlab.com/freedesktop-sdk/freedesktop-sdk

**Comment by TingPing** [comment]
@alexander255 I'm entirely aware of what they do but that doesn't make it less of a hack. The concern with modules is that can do arbitrary things and potentially cause problems and the big selling point of Flatpak is reproducablity and portability, injecting modules is counter to that. The same concern somewhat applies to themes but at least those aren't fully arbitrary code.

**Comment by matthiasclasen** [comment]
I am very much against an extension point for gtk modules.

**Comment by alexander255** [comment]
@matthiasclasen: So what would your alternative suggestion be to make GTK support global menus?
(Just not supporting it is a very lame alternative, since the global menu is quite popular on KDE for instance.)

**Comment by matthiasclasen** [comment]
gtk supports global menus, using gmenumodel.
The screenscaping done behind the scenes by those modules is not supportable to begin with.

**Comment by alexander255** [comment]
> gtk supports global menus, using gmenumodel.

As I stated just two hours before GMenuModel is a horrible „replacement“ for GtkMenuBar. It works if all you want to export is a “hamburger button” with 3 – 10 options, but for complex applications with many different options its just not a usable alternative. I know, GNOME devs decided for all of us that nobody actually needs anything resembling a more classical menu bar and that all applications having such a thing should just “make it work with less” – it's what they did to their own application stack; doesn't look like anybody outside of the GNOME project agrees however (KDE, Mate & XFCE certainly do not), so just “letting GtkMenuBar die” is not a proper solution. (Remember how the G in GTK stands for GIMP/GNU, not GNOME?!)

I'm sorry for repeating this whole debate again, but pretending that the problem isn't there won't make it go away…

> The screenscaping done behind the scenes by those modules is not supportable to begin with.

Fully agreement on that however. So what we should actually do is make GTK 3 & 4 export proper menus themselves, no?

**Comment by matthiasclasen** [comment]
I don't think the debate is worth having. It doesn't matter if you like menu models or not. They are there, and they are the only form of global menus that you will get from gtk.

**Comment by alexander255** [comment]
> It doesn't matter if you like menu models or not. They are there, and they are the only form of global menus that you will get from gtk.

I tried to point out at every occasion that there is plenty of evidence that this is not just my private special PoV, but something that a lot of people want (*Global* Menu*Bar*s, I mean).

> I don't think the debate is worth having.

*TL;DR* on our conversation: Exporting GtkMenuBars from GTK will not happen and we'll also try to block anybody who tries to get back this feature using a GTK model, since that is a horrible way of doing that (likely true) – we won't give you any alternative however.

Did I misinterpret/misrepresent anything there?

**Comment by TingPing** [comment]
Exporting GtkMenu is an awful task, have you read dbusmenu-gtk? Its a nightmare.

You are probably better off making small improvements to GMenuModel to fit your needs but that discussion needs to happen in GLib/GTK not here.

**Comment by alexander255** [comment]
@TingPing: I may have spoken too soon (and too much). Reading up on this stuff it turns out that one *can* actually create classic menu bars  of arbitrary complexity and not just single-depth “hamburger buttons” from GMenuModel, yes?

In that case I extend my sincere apologies to @matthiasclasen for bothering you on this, in my mind any time an application switched to GMenuModel it meant that it also switched to “hamburger button”-style menu layout (probably cause they were all GNOME apps?), so I wasn't aware that one does not imply the other.

**Comment by TingPing** [comment]
> Reading up on this stuff it turns out that one can actually create classic menu bars of arbitrary complexity and not just single-depth “hamburger buttons” from GMenuModel, yes?

Correct.

**Comment by TingPing** [comment]
Anyway I'll close this since extensions are defined in the runtimes not by flatpak, so you would request then there (even if this one may be controversial to some).
